### PR TITLE
Disable NVIDIA's threaded OpenGL optimization on Windows

### LINKED ISSUE #33969: Single frame spikes even in an empty project
**Godot version:**
3.1.1, 3.2b2

**OS/device including version:**
Windows 10 Pro 1909 x64

**Issue description:**
Every ~10 seconds i get a one-frame spike followed by 1-3 really short frames. At first i thought it was my fault, but problem persists even in an empty project.

Exported project exhibits same behavior.

My specs are i5 7300, 16gb DDR4, GTX 1050

Profiler screenshots:

Frame before the spike:
Annotation 2019-11-28 204133

Spike:
Annotation 2019-11-28 204205

Frame after the spike:
Annotation 2019-11-28 204229

**Steps to reproduce:**
1) Attach a script with empty "_process" method anywhere in the scene, so that Editor may pause the scene (only for a profiler screenshot)
2) Run project and wait for ~10 sec

**EDIT:**
So, in trying to narrow down the issue(at least as far as my abilities go), i tested this on some other machines

i7 - 9750H / 16 Gb DDR4 / RTX-2060  w/ Windows 10 - same thing,
i3 - 3240 / 4 Gb RAM (didn't check at the time, i believe DDR3) / GeForce 8600GT w/ Windows 8.1 - same thing

I tried to check the first machine (i5 7300, 16gb DDR4, GTX 1050) with Manjaro 18.1.5 and didn't notice any of similar problems (although i couldn't get Vsync to work). Project, exported on Manjaro to Windows exhibits same behavior

3.2 rc4 didn't help.

**EDIT 2: The Editing**
I decided to try Vulkan branch just to see if something different happens.
Spikes aren't there, but i noticed really strange things happen with everything that is normalized for time (with delta), i haven't noticed this before because of short lag, but in Vulkan that short lag isn't there, and after some investigating, i found that sometimes delta is negative(with same frequency as spikes in stable version). Going back to 3.2, negative delta does happen at the same time as spikes, but not everytime.

Annotation 2020-02-06 222053

So i guess, question is, could VSync issues and negative delta be related?

### LINKED ISSUE #45660: nVidia "Threaded Optimization" Driver Setting Intefering with Godot Preload Capabilities
**Godot version:** 3.2.3.stable.official (Steam Release)

**OS/device including version:**
Windows 10 Pro 19041.423
CPU: AMD FX-8350 4 GHz
GPU: GeForce GTX 1070 - 8 GB VRAM
RAM: 16 GB
Using an HDD, not SSD, for Storage

**Issue description:**
Just a couple days ago I made a report on an issue I was having with the particle system which didn't seem related to shader caching. I reported it here: https://github.com/godotengine/godot/issues/45613 . It is important to keep in mind while I explain the following issue that it seems the particle issue I reported may or may not be related, it's actually really hard to tell.

Anyways, after some investigation I discovered I was getting hiccups with instancing literally ANYTHING which was going to render a shader of any kind, it was just much more noticeable with particles for whatever reason. These hiccups were only occurring the first time an instanced object would become visible in a scene, so naturally the thought would be a caching issue and that stuff would need to be preloaded... except... I AM preloading things, and if I run my projects in a window instead of full screen, these hiccups disappear. Some mysterious difference was occurring between running my projects in a window versus full screen which was causing preloading/caching to fail.

After MANY hours of investigation, I have discovered something insidious: nVidia GeForce drivers have a setting called "Threaded Optimization", normally set to "Auto" which is used to spread CPU usage by hardware accelerated games over multiple cores. 99% of the time, changing this has zero effect on the performance of a game, though it often changes how the CPU usage for a game is spread across a multi-core processor.

...however, if I force "Threaded Optimization" into an "Off" state in my driver settings, the hiccups I was having with first-time instancing of supposedly preloaded objects while running my Godot projects full screen disappears!

So... there seems to be some sort of incompatibility with preloading assets in Godot, while running a Godot app full screen, on a GeForce card, with the card's Threaded Optimization driver setting set to on or auto, keeping in mind, it's set to "Auto" by default.

...I don't have a clue how this is supposed to be addressed by the Godot devs, since the problem is a driver setting. If this was just a single game you'd just contact nVidia and say, "Hey, could you include a profile which disables Threaded Optimization for our game? Thanks!" ...but for a game ENGINE, which makes multiple games, I really don't know what the solution here is, other than to instruct players how to disable this setting manually. :/

**Steps to reproduce:**
1. This must be done on a system with a GeForce graphics chipset.
2. Go into the nVidia driver settings and set the global profile's "Threaded Optimization" setting to "Auto" or "On". It should already be "Auto" by default.
3.  Ensure that Godot Engine itself and any projects made with it are NOT in the list of settings for individual programs in the nVidia driver settings.
4. Open any Godot project which has the ability to instance objects on command and run it full screen.
5. The very first time a shader appears on screen when instancing an object, there will be a very slight hiccup in the framerate. Subsequent instances of the object will not produce this skip in the framerate.
6. Go back into the editor and switch the project to run in a window. The hiccups will no longer occur.
7. Go back into the editor and go back to running full screen, but also go into the nVidia driver settings again and set "Threaded Optimization" to "Off". Run the project and the hiccups remain gone.
8. Turn Threaded Optimization back to "On" or "Auto" and the hiccups return.

**Minimal reproduction project:**
Literally any project which instances objects with unique shaders applied to them and is run full screen will trigger this issue.

### PR DESCRIPTION
**Pull request by EIREXE**
Disables NVIDIA's threaded optimization which causes random stuttering on Windows with NVIDIA GPUs, virtually all performant OpenGL programs (like ryujinx) disable this to prevent stuttering and it's a well documented issue.

Should fix #33969

- *Production edit: This closes https://github.com/godotengine/godot/issues/45660.*

### CODE REVIEW

File: thirdparty/README.md [hunk 1]
```diff
@@ -515,6 +515,17 @@ Files extracted from the upstream source:
 - `LICENSE.txt` and `CHANGELOG.md`
 
 
+## NVAPI
```
Comment (bruvzg) [inline thirdparty/README.md:518]: Third-party lib info should be added to the `/COPYRIGHT.txt` as well.

File: platform/windows/gl_manager_windows.cpp [hunk 2]
```diff
@@ -66,6 +68,166 @@ static String format_error_message(DWORD id) {
...
+		return;
+	}
+
+	String app_executable_name = OS::get_singleton()->get_executable_path().get_file();
+	String app_friendly_name = GLOBAL_GET("application/config/name");
+	String app_profile_name = app_friendly_name + " Nvidia Profile";
+
+	NvDRSProfileHandle profile_handle;
+	NvAPI_UnicodeString nvapi_profile_name;
+	memcpy(nvapi_profile_name, app_profile_name.utf8().get_data(), sizeof(wchar_t) * (app_profile_name.utf8().size()));
```
Comment (bruvzg) [inline platform/windows/gl_manager_windows.cpp:163]: `NvAPI_UnicodeString` seems to be UTF-16, so this is wrong, should be:

```suggestion
	NvAPI_UnicodeString nvapi_profile_name;
	Char16String app_profile_name_u16 = app_profile_name.utf16();
	memcpy(nvapi_profile_name, app_profile_name_u16.get_data(), app_profile_name_u16.size());
```

Or you can probably use `(NvAPI_LPCWSTR)(app_profile_name.utf16().ptrw())` instead of `nvapi_profile_name` directly.

File: platform/windows/gl_manager_windows.cpp [hunk 3]
```diff
@@ -66,6 +68,166 @@ static String format_error_message(DWORD id) {
...
+
+	int status = NvAPI_DRS_FindProfileByName(session_handle, nvapi_profile_name, &profile_handle);
+
+	if (status != 0) {
+		print_verbose("NVAPI: Profile not found, creating....");
+
+		NVDRS_PROFILE profile_info;
+		profile_info.version = NVDRS_PROFILE_VER;
+		profile_info.isPredefined = 0;
+		memcpy(profile_info.profileName, app_profile_name.utf8().get_data(), sizeof(wchar_t) * (app_profile_name.utf8().size()));
```
Comment (bruvzg) [inline platform/windows/gl_manager_windows.cpp:173]: Same here.

File: platform/windows/gl_manager_windows.cpp [hunk 4]
```diff
@@ -66,6 +68,166 @@ static String format_error_message(DWORD id) {
...
+			return;
+		}
+
+		NVDRS_APPLICATION_V4 app;
+		app.version = NVDRS_APPLICATION_VER_V4;
+		app.isPredefined = 0;
+		app.isMetro = 1;
+		app.isCommandLine = 1;
+		memcpy(app.appName, app_executable_name.utf8().get_data(), sizeof(wchar_t) * (app_executable_name.utf8().size()));
+		memcpy(app.userFriendlyName, app_friendly_name.utf8().get_data(), sizeof(wchar_t) * (app_friendly_name.utf8().size()));
```
Comment (bruvzg) [inline platform/windows/gl_manager_windows.cpp:187]: Same here.

File: platform/windows/gl_manager_windows.cpp [hunk 5]
```diff
@@ -66,6 +68,166 @@ static String format_error_message(DWORD id) {
...
+void GLManager_Windows::_nvapi_disable_threaded_optimization() {
+	HMODULE nvapi = 0;
+#ifdef _WIN64
+	nvapi = LoadLibraryA("nvapi64.dll");
+#else
+	nvapi = LoadLibraryA("nvapi.dll");
+#endif
+
+	if (nvapi == NULL) {
+		print_verbose("Error loading NVAPI library");
```
Comment (bruvzg) [inline platform/windows/gl_manager_windows.cpp:112]: Probably should not print anything, to avoid errors on non NVIDIA hardware.

**Comment by EIREXE** [comment]
@bruvzg Should be OK now, I also added a fallback profile name for use in the Project Manager

**Comment by clayjohn** [comment]
Seems like a shame to add 25,000 lines of code to disable a driver setting only on one brand of GPUs, but looking at the build artifacts, overall size has changed by less than 1 mb (the size for windows build artifacts in this PR are reported at the same size as master) so maybe it isn't a big deal

**Comment by aaronfranke** [comment]
Why does disabling a driver setting require 25k lines of third-party code?

**Comment by EIREXE** [comment]
> Why does disabling a driver setting require 25k lines of third-party code?

ask nvidia, although we could just copy some nvidia types over and use only those

**Comment by EIREXE** [comment]
> @EIREXE would it be possible (aka without too much trouble) to trim everything that isn't the threaded optimization part from the thirdparty code or is it too jumbled up for that?
> 
> I can't imagine that most big graphics programs just import this huge gigantic 25k lines header for this idiotic problem, it's silly.
> 
> I swear most weird graphics problems we stumble upon here are NVIDIA's fault.
> 
> Edit: obv I misread your comment and you already commented it as a possibility, sorry. Obviously try doing this if it's humanly feasible, knowing nvidia I wouldn't be surprised if it were a jumbled mess.

I've done just that, seems to work fine.

**Comment by clayjohn** [comment]
@EIREXE I guess you can remove the original NVAPI file right?

**Comment by EIREXE** [comment]
> @EIREXE I guess you can remove the original NVAPI file right?

oh my b, I forgot to remove it

File: doc/classes/ProjectSettings.xml [hunk 6]
```diff
@@ -1998,6 +1998,10 @@
 		<member name="rendering/gl_compatibility/item_buffer_size" type="int" setter="" getter="" default="16384">
 			Maximum number of canvas items commands that can be drawn in a single viewport update. If more render commands are issued they will be ignored. Decreasing this limit may improve performance on bandwidth limited devices. Increase this limit if you find that not all objects are being drawn in a frame.
 		</member>
+		<member name="rendering/gl_compatibility/nvidia_disable_threaded_optimization" type="bool" setter="" getter="" default="true">
+			If [code]true[/code] disables the threaded optimization feature from the NVIDIA drivers, which are known to cause stuttering on most OpenGL applications.
+			[b]Note:[/b] This setting only works on Windows, as threaded optimization is disabled by default on other platforms.
```
Comment (Calinou) [inline doc/classes/ProjectSettings.xml:2003]:
```suggestion
			If [code]true[/code], disables the threaded optimization feature from the NVIDIA drivers, which are known to cause stuttering in most OpenGL applications.
			[b]Note:[/b] This setting only works on Windows, as threaded optimization is disabled by default on other platforms.
```

**Review by Calinou — APPROVED** [review]

**Comment by clayjohn** [comment]
> > @EIREXE I guess you can remove the original NVAPI file right?
> 
> oh my b, I forgot to remove it

@EIREXE It looks like you still haven't pushed the updated version, the diff still adds 25,000 lines of code. 

No pressure, but when you get a chance could you push your changes?

**Comment by EIREXE** [comment]
> > > @EIREXE I guess you can remove the original NVAPI file right?
> > 
> > 
> > oh my b, I forgot to remove it
> 
> @EIREXE It looks like you still haven't pushed the updated version, the diff still adds 25,000 lines of code.
> 
> No pressure, but when you get a chance could you push your changes?

Oh I thought I did, give me a second

File: COPYRIGHT.txt [hunk 7]
```diff
@@ -1939,3 +1939,8 @@ License: Zlib
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.
+
+Files: ./thirdparty/nvapi/nvapi_minimal.h
+Comment: Stripped down version of "nvapi.h" from the NVIDIA NVAPI SDK
+Copyright: 2019-2022, NVIDIA CORPORATION. All rights reserved.
+License: MIT
```
Comment (clayjohn) [inline COPYRIGHT.txt:1946]: This should be up between msdfgen and oidn as it is in the thirdparty readme. Other than that, this PR seems ready to merge!

Comment (akien-mga) [inline COPYRIGHT.txt:1946]:
```suggestion
Copyright: 2019-2022, NVIDIA Corporation
License: Expat
```
(This file follows a format defined by Debian and they use "Expat" for the typical MIT license, to disambiguate from a few other licenses originating from MIT with slightly different languages.

File: doc/classes/ProjectSettings.xml [hunk 8]
```diff
@@ -1998,6 +1998,10 @@
 		<member name="rendering/gl_compatibility/item_buffer_size" type="int" setter="" getter="" default="16384">
 			Maximum number of canvas items commands that can be drawn in a single viewport update. If more render commands are issued they will be ignored. Decreasing this limit may improve performance on bandwidth limited devices. Increase this limit if you find that not all objects are being drawn in a frame.
 		</member>
+		<member name="rendering/gl_compatibility/nvidia_disable_threaded_optimization" type="bool" setter="" getter="" default="true">
+			If [code]true[/code] disables the threaded optimization feature from the NVIDIA drivers, which are known to cause stuttering on most OpenGL applications.
```
Comment (akien-mga) [inline doc/classes/ProjectSettings.xml:2002]:
```suggestion
			If [code]true[/code], disables the threaded optimization feature from the NVIDIA drivers, which are known to cause stuttering on most OpenGL applications.
```

File: platform/windows/gl_manager_windows.cpp [hunk 9]
```diff
@@ -33,11 +33,13 @@
 #ifdef WINDOWS_ENABLED
 #ifdef GLES3_ENABLED
 
+#include "core/config/project_settings.h"
+#include "core/version.h"
+#include "thirdparty/nvapi/nvapi_minimal.h"
+#include <dwmapi.h>
 #include <stdio.h>
 #include <stdlib.h>
```
Comment (akien-mga) [inline platform/windows/gl_manager_windows.cpp:41]:
```suggestion
#include "core/config/project_settings.h"
#include "core/version.h"

#include "thirdparty/nvapi/nvapi_minimal.h"

#include <dwmapi.h>
#include <stdio.h>
#include <stdlib.h>
```

File: thirdparty/README.md [hunk 10]
```diff
@@ -515,6 +515,15 @@ Files extracted from the upstream source:
...
```
Comment (akien-mga) [inline thirdparty/README.md:518]:
```suggestion
## nvapi
```
These section titles match the actual folder names.

File: thirdparty/nvapi/nvapi_minimal.h [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+     NvU32                      numOfSettings;      //!< Total number of settings applied for this Profile. Read-only
+} NVDRS_PROFILE_V1;
+
+typedef NVDRS_PROFILE_V1         NVDRS_PROFILE;
+
+//! Macro for constructing the version field of ::NVDRS_PROFILE
+#define NVDRS_PROFILE_VER1       MAKE_NVAPI_VERSION(NVDRS_PROFILE_V1,1)
+#define NVDRS_PROFILE_VER        NVDRS_PROFILE_VER1
+
+#endif
```
Comment (akien-mga) [inline thirdparty/nvapi/nvapi_minimal.h:175]: Missing newline at EOF.

**Comment by EIREXE** [comment]
Sorry for the many mess-ups, should be fine now I hope.

**Review by akien-mga — APPROVED** [review]
Looks great!

**Comment by akien-mga** [comment]
Thanks!

**Comment by Calinou** [comment]
Should we also set `__GL_THREADED_OPTIMIZATIONS` to `0` on Linux before creating the OpenGL context? See https://download.nvidia.com/XFree86/Linux-x86_64/435.17/README/openglenvvariables.html#:~:text=or%0A~%2F.profile.-,threaded%20optimizations,-The%20NVIDIA%20OpenGL.

File: platform/windows/gl_manager_windows.cpp [hunk 12]
```diff
@@ -64,6 +69,171 @@ static String format_error_message(DWORD id) {
...
+			NvAPI_GetErrorMessage__(status, err_desc);
+			print_verbose(vformat("%s: %s(code %d)", msg, err_desc, status));
+		}
+		return false;
+	}
+	return true;
+}
+
+// On windows we have to disable threaded optimization when using NVIDIA graphics cards
+// to avoid stuttering, see https://github.com/microsoft/vscode-cpptools/issues/6592
```
Comment (jsjtxietian) [inline platform/windows/gl_manager_windows.cpp:102]: The link to vscode's issue seems irrelevant, am I understand it wrong?

Comment (EIREXE) [reply platform/windows/gl_manager_windows.cpp:102]: I am stupid indeed, the correct link should be https://stackoverflow.com/questions/36959508/nvidia-graphics-driver-causing-noticeable-frame-stuttering/37632948

Comment (jsjtxietian) [reply platform/windows/gl_manager_windows.cpp:102]: Thanks !

Comment (EIREXE) [reply platform/windows/gl_manager_windows.cpp:102]: I'm very busy lately, so I would be very thankful if you could do it for me

File: platform/windows/gl_manager_windows.cpp [hunk 13]
```diff
@@ -64,6 +69,171 @@ static String format_error_message(DWORD id) {
...
+
+	NVDRS_SETTING setting;
+	setting.version = NVDRS_SETTING_VER;
+	setting.settingId = OGL_THREAD_CONTROL_ID;
+	setting.settingType = NVDRS_DWORD_TYPE;
+	setting.settingLocation = NVDRS_CURRENT_PROFILE_LOCATION;
+	setting.isCurrentPredefined = 0;
+	setting.isPredefinedValid = 0;
+	int thread_control_val = OGL_THREAD_CONTROL_DISABLE;
+	if (!GLOBAL_GET("rendering/gl_compatibility/nvidia_disable_threaded_optimization")) {
```
Comment (jsjtxietian) [inline platform/windows/gl_manager_windows.cpp:212]: Can we check `nvidia_disable_threaded_optimization` at the beginning of the function to avoid crash in #84880 ?
